libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -116,37 +374,168 @@ proc rebalanceMesh(g: GossipSub, topic: string) {.async.} =
   # create a mesh topic that we're subscribing to
 
   var
-    grafts, prunes: seq[PubSubPeer]
+    grafts, prunes, grafting: seq[PubSubPeer]
 
-  if g.mesh.peers(topic) < GossipSubDlo:
+  let npeers = g.mesh.peers(topic)
+  if npeers  < GossipSubDlo:
     trace "replenishing mesh", peers = g.mesh.peers(topic)
     # replenish the mesh if we're below Dlo
     grafts = toSeq(
       g.gossipsub.getOrDefault(topic, initHashSet[PubSubPeer]()) -
       g.mesh.getOrDefault(topic, initHashSet[PubSubPeer]())
     )
 
-    shuffle(grafts)
+    grafts.keepIf do (x: PubSubPeer) -> bool:
+      # avoid negative score peers
+      x.score >= 0.0 and
+      # don't pick explicit peers
+      x.peerId notin g.parameters.directPeers and
+      # and avoid peers we are backing off
+      x.peerId notin g.backingOff
+
+    # sort peers by score
+    grafts.sort do (x, y: PubSubPeer) -> int:
+      let
+        peerx = x.score
+        peery = y.score
+      if peerx < peery: -1
+      elif peerx == peery: 0
+      else: 1
 
     # Graft peers so we reach a count of D
     grafts.setLen(min(grafts.len, GossipSubD - g.mesh.peers(topic)))
 
     trace "grafting", grafts = grafts.len
+    for peer in grafts:
+      if g.mesh.addPeer(topic, peer):
+        g.grafted(peer, topic)
+        g.fanout.removePeer(topic, peer)
+        grafting &= peer
+    
+  elif npeers < g.parameters.dOut:
+    trace "replenishing mesh outbound quota", peers = g.mesh.peers(topic)
+    # replenish the mesh if we're below Dlo
+    grafts = toSeq(
+      g.gossipsub.getOrDefault(topic, initHashSet[PubSubPeer]()) -
+      g.mesh.getOrDefault(topic, initHashSet[PubSubPeer]())
+    )
+
+    grafts.keepIf do (x: PubSubPeer) -> bool:
+      # get only outbound ones
+      x.outbound and
+      # avoid negative score peers
+      x.score >= 0.0 and
+      # don't pick explicit peers
+      x.peerId notin g.parameters.directPeers and
+      # and avoid peers we are backing off
+      x.peerId notin g.backingOff
+
+    # sort peers by score
+    grafts.sort do (x, y: PubSubPeer) -> int:
+      let
+        peerx = x.score
+        peery = y.score
+      if peerx < peery: -1
+      elif peerx == peery: 0
+      else: 1
+
+    # Graft peers so we reach a count of D
+    grafts.setLen(min(grafts.len, g.parameters.dOut - g.mesh.peers(topic)))
+
+    trace "grafting outbound peers", topic, peers = grafts.len
 
     for peer in grafts:
       if g.mesh.addPeer(topic, peer):
+        g.grafted(peer, topic)
         g.fanout.removePeer(topic, peer)
+        grafting &= peer
+    
 
   if g.mesh.peers(topic) > GossipSubDhi:
     # prune peers if we've gone over Dhi
     prunes = toSeq(g.mesh[topic])
-    shuffle(prunes)
-    prunes.setLen(prunes.len - GossipSubD) # .. down to D peers
-
-    trace "pruning", prunes = prunes.len
-    for peer in prunes:
-      g.mesh.removePeer(topic, peer)
 
+    # sort peers by score (inverted)
+    prunes.sort do (x, y: PubSubPeer) -> int:
```

Comment:
`proc byScore(x,y: PSP): int = x.score-y.score`? then `prunes.sort(byScore)`

Reply:
Ok, how about:
`peers.sort do (x, y: PubSubPeer) -> int: (x.score - y.score).int` ?

Reply:
or you think the closure might be "slower"?

Reply:
bah the way you propose anyway is cleaner.. let's do that!

Reply:
fixed